pennylane/capture/autograph/ag_primitives.py

```diff
@@ -131,6 +135,157 @@ def assert_iteration_inputs(inputs, symbol_names):
             ) from e
 
 
+def assert_iteration_results(inputs, outputs, symbol_names):
+    """The results of a for loop should have the identical type as the inputs, since they are
+    "passed" as inputs to the next iteration. A mismatch here may indicate that a loop carried
+    variable was initialized with wrong type.
+    """
+
+    for i, (inp, out) in enumerate(zip(inputs, outputs)):
+        inp_t, out_t = jax.api_util.shaped_abstractify(inp), jax.api_util.shaped_abstractify(out)
+        if inp_t.dtype != out_t.dtype or inp_t.shape != out_t.shape:
+            raise AutoGraphError(
+                f"The variable '{symbol_names[i]}' was initialized with the wrong type, or you may "
+                f"be trying to change its type from one iteration to the next. "
+                f"Expected: {out_t}, Got: {inp_t}"
+            )
+
+
+# pylint: disable=too-many-positional-arguments
+def _call_pennylane_for(
+    start,
+    stop,
+    step,
+    body_fn,
+    get_state,
+    set_state,
+    symbol_names,
+    enum_start=None,
+    array_iterable=None,
+):
+    """Dispatch to a PennyLane implementation of for loops."""
+
+    # Ensure iteration arguments are properly initialized. We cannot process uninitialized
+    # loop carried values as we need their type information for tracing.
+    init_iter_args = get_state()
+    assert_iteration_inputs(init_iter_args, symbol_names)
+
+    @qml.for_loop(start, stop, step)
+    def functional_for(i, *iter_args):
+        # Assign tracers to the iteration variables identified by AutoGraph (iter_args in mlir).
+        set_state(iter_args)
+
+        # The iteration index/element (for <...> in) is already handled by the body function, e.g.:
+        #   def body_fn(itr):
+        #     i, x = itr
+        #     ...
+        if enum_start is None and array_iterable is None:
+            # for i in range(..)
+            body_fn(i)
```

Comment:
So what is the actual signature of `body_fn`? Does it accept a tuple or an integer or an object?

Reply:
So, `malt`/`autograph` do this part, not our implementation, so take any details here about what happens under the hood with a big grain of salt. But my understanding of what autograph is passing to the `for_stmt` function is that `body_fn` is what to do inside the loop, i.e. in the loop:

```
for x in range(3):
    qml.X(x, 0)
```
    
we have the first section of the if-statement, and the body function is `lambda i: qml.X(i, 0)`. It could also be something like:

```
params = [0.1, 0.2, 0.3]
for (i, x) in enumerate(array):
    params[i] += x
```

then we are in the `else` clause, and the body function is 

```
def body_fn(i, x): 
    params[i] += x
```

it has access to some `state` where it knows all the initialized variables, so it can update `params`, even though params isn't passed to the body function (but this is why its important everything is initialized and types don't change). Then in the end, it gets the `state` of all the initialized variables after the loop is applied, and updates all their current values.

The `body_fn` is generated by autograph from parsing the user-code, and it always takes either an index, an array element, or both, where which one depends on the syntax the user is creating a for-loop with.